### ISSUE TITLE
Call To Arms: Documentation

### ISSUE THREAD

**Issue by mcdonnelldean** [rjrodger/seneca#155]
Hi everyone,

So as you may know, documentation for Seneca has been somewhat lacking. We want to change this and we are looking for your help. There are lots of things we want to do and there are plenty of ways to help out. Below I'll loosely group all of the work I can see that needs attention.

### UPDATE 2:
We are now ready to accept contributions to the new docs site. Come join us at https://github.com/senecajs/senecajs.org and say hi! 

I'll be squashing and moving this thread over to the new repo tonight, see ya'll on the other side :dancer: 

### UPDATE:
We are currently squirrelling away getting docs on a repo ready for public consumption. Once this is live it will have a dedicated issue's list to work off and we can start breaking out tasks better. It may take a day or two longer than expected. Once this is ready I'll update with a link to the repo.

## Site Change Over
The current doc site isn't very mobile friendly and needs a general refresh. Thankfully a template and theme have already been contributed. On top of this, work was done on to move the site to [Metalsmith][], improve offline usage, and simplify contributor experience. 

My plan is to have a branch live with this version of the site in the next day or two. It will be this branch that people will contribute to until we reach a release point, at which stage we will mostly likely revert back to contributing to the main branch.

Once this is done, if you are a web person we would really appreciate help on slimming the overall site down. There are plenty of places we could optimise to improve load times and ease of use. If you feel you can contribute here, @ me directly.

## Big Documentation
If you are familiar with Seneca, or like the idea of writing bigger more involved pieces of writing, we have a number of topics that could use some good docs. Below I have added a list of priority topics that need love and attention:

- API Reference
- Options Reference Article
- Writing a Plugin
- Using Seneca with Rest
- Data Entities Dive
- Logging Deep Dive
- Commands Deep Dive
- Guides on various Plugins

The list above is not exhaustive, it is more representative of priority. If you feel you have some big documentation, guide or other similar sized works to contribute, please talk to us.

## Small Documentation
If big docs aren't your thing we have plenty of smaller chunks of work to be done. How-to's in particular are high on users wish lists.

- Improve ReadMe content for existing plugins
- Add support ReadMe's for existing examples
- How to guides for various Transports and Stores.

## Other Work
There is always work to be done that is not just simple documentation, any work in the following areas would be greatly appreciated too.

- Adding examples to [Seneca Examples][]
- Documenting and refactoring current examples
- Ensuring sample code is up to date
- Proof reading for Errata in content and code.
- Collecting potential topics requiring documentation

## For the Adventurer
If you are feeling particularly adventurous we have just the project for you. We would like to see the creation of a Metalsmith Plugin that could scan npm / github for plugins and generate a page with this
data. This would function exactly like our plugins page. We would essentially like to replace the manual
effort required to keep this updated..... Are you feeling Adventurous :smile:  

## FAQ
Q: Who decides what I work on?
A: You Do! Just sound out that you would like to tackle a piece of work and get to it.

Q: My English isn't that great, can I still help?
A: Absolutely, there are plenty of people that can proofread your work, don't sweat it!

Q: What if I want to just contribute code examples?
A: We are looking for good sample code to, see [Seneca Examples]

Q: Can I write a guide for _insert plugin here_?
A: Please do! The more the merrier

Q: How will PR's be handled
A: Once the branch is created I will add detailed instructions on how to contribute new work.

---

Questions, Thoughts, Hugs, Comments?

[Seneca Examples]: https://github.com/rjrodger/seneca-examples
[Metalsmith]: http://www.metalsmith.io/

**Comment by psnider** [comment]
I'd like to help with the API docs, as I think it will help me learn seneca better.
Also, I'd like to create a TypeScript declarations file for seneca, and I'll need good docs before I can do this!

**Comment by mcdonnelldean** [comment]
@psnider Awesome, as soon as I have these available I'll ping you, they are mostly complete, they just need a little attention and while learning is the perfect time to add this I think.

**Comment by genexp** [comment]
We'd like to help too.  I feel that Seneca is where Ember was a few years ago.  With some good documentation, and some of the changes in the queue for 0.7 (multiple parameters, and maybe promises) this should be the main choice for any service developer using node.

In addition to basic commands, I'd like to see some example patters.  For example, I'd like to see a CQRS with Event Sourcing example, a simple HTTP service example (with Express and Hapi) etc.

**Comment by mcdonnelldean** [comment]
@genexp Awesome!

I'll add your suggestions to the list too. I'm a big fan of Event Sourcing myself so I would love to see docs on both topics!

**Comment by blainsmith** [comment]
Yeah there are some interesting cases where if you use `.sub()`, `.act()`, and `.add` appropriately you can get pretty far.

**Comment by genexp** [comment]
hey @kixxauth - You should be on this thread.  So should @brkattk
<https://github.com/brkattk>

**Comment by psnider** [comment]
Here's a lame first pass at a declarations file. 
It needs more work!
This is just to give an idea, and this is not correct yet...
... and it will be much larger!

```
declare module "seneca" {

    interface Options {
        log?: any;
    }

    function s(options: Options): s.Seneca;


    // what does seneca do about logging?
    function log(): void;


    module s {

         interface PluginOptions {
         }

         interface ClientOptions {
         }

         interface ListenOptions {
         }

        type Pattern = string | any;
        type AddCallback = (error: Error, what: any) => void;
        type ActCallback = (error: Error, result: any) => void;
        type CloseCallback = (optional: any, done: (error: Error) => void) => void;

        interface Seneca {

            version: string;

            (options: Options): s.Seneca;
            error(done: () => void): void;
            on(eventName: string, callback: (error: Error) => void);
            close()
            use(pluginName: string): void;
            use(quickplugin: (opts) => void): void;
            client(options: ClientOptions): void;
            listen(options: ListenOptions): void;

            ready(callback: (error: Error) => void);

            add(pattern: Pattern, action: AddCallback): void;
            add(pattern: Pattern, paramspec: any, action: AddCallback): void;

            act(pattern: Pattern, respond: ActCallback);

            make(entityType: string, entity: any): void;

            // @param name reference to plugin
            export(name: string): void;

            pin(pattern: Pattern): void;
        }
    }

    export = s;

}
```

**Comment by mcdonnelldean** [comment]
Awesome! Can I get you to create a new issue with this? It deserves it's own, not to be lumped in with documentation. There may be people already interested in TypeScript that could help out too.

**Comment by mcdonnelldean** [comment]
Closed as we now have a dedicated repo
